new install_jupyter_kernel script for simplified DR installations

sbailey opened the pull request:
This PR provides `install_jupyter_kernel` as a proposed replacement for `install_jupyter_kernel.sh`, with the primary motivation of dramatically simplifying our [current multi-step process](https://github.com/desihub/tutorials/blob/main/getting_started/installing_EDR_kernels_at_NERSC.md) for installing kernels for a data release environment, which also currently requires us to maintain pre-built kernel files for every software/datarelease combination.

The idea is that we would provide a link to this script in /global/common/software/desi/, and then the instructions would become:

For DESI users to install a kernel for a software release
```
/global/common/software/desi/install_jupyter_kernel -v 24.11
```

For non-DESI NERSC users *or* DESI users to install a kernel for a software release and switch to a data release environment
```
/global/common/software/desi/install_jupyter_kernel -v 24.11 -r dr1
```

Like the existing `install_jupyter_kernel.sh` script, it derives which activation script to use based upon $SHELL, and it generates the kernel file on the fly instead of requiring the user to pick a specific pre-built kernel.

The script is python 2.7 compatible, so it works with the ancient perlmutter default system python without requiring pre-loading a DESI-environment.  I've also tested it with python 3.10.14.  The script also has an --outdir option to write to somewhere other than $HOME/.local/share/jupyter/kernels for testing.

Example using my current checkout:
```
[perlmutter] $CFS/desi/users/sjbailey/code/desimodules/install_jupyter_kernel -v 23.1 -r edr
Wrote /global/homes/s/sjbailey/.local/share/jupyter/kernels/desi-23.1-edr/kernel.json
[perlmutter] cat /global/homes/s/sjbailey/.local/share/jupyter/kernels/desi-23.1-edr/kernel.json
{
 "language": "python",
 "argv": [
  "/global/common/software/desi/activate_desi_jupyter.sh",
  "23.1", "edr",
  "{connection_file}"
 ],
 "display_name": "DESI 23.1 edr"
}
```
Comparing to the existing pre-build edr-23.1 kernel:
```
[perlmutter] cat /global/common/software/desi/kernels/desi-edr-23.1-bash/kernel.json
{
 "language": "python",
 "argv": [
  "/global/common/software/desi/activate_desi_jupyter.sh",
  "23.1",
  "edr",
  "{connection_file}"
 ],
 "display_name": "DESI EDR 23.1"
}
```

Swapping the order of software release version vs. data release version was purposeful so that optional part of the name (the data release name) is at the end instead of in the middle.  Functionally they are the same.

For the moment the display name uses the lowercase "edr" or "dr1" to match the desitree module version that will be swapped under the hood, but I wouldn't be opposed to making the uppercase for the display name.  We might also consider forcing the `-r` option to lowercase since that is what we use for desitree, so that this would also work even with the "incorrect" capitalization:
```
install_jupyter_kernel -v 24.11 -r DR1
```

weaverba137 commented on install_jupyter_kernel:
```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env python
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+from __future__ import print_function
```
This is not needed.

sbailey replied:
Removed

weaverba137 reviewed (changes requested):
I have one minor, almost stylistic, suggestion.

weaverba137 commented:
> Swapping the order of software release version vs. data release version was purposeful so that optional part of the name (the data release name) is at the end instead of in the middle. Functionally they are the same.

If you actually did change the order of the inputs, that would require a change to `activate_desi_juypter.(sh|csh)`. But it looks like you changed the *formatting* of the inputs, not the *order*. Is that what you meant?

weaverba137 commented:
> dramatically simplifying our [current multi-step process](https://github.com/desihub/tutorials/blob/main/getting_started/installing_EDR_kernels_at_NERSC.md)

Indeed that's a lot of steps. I went back and reviewed that. Where it says:

> Access https://jupyter.nersc.gov/hub/login and initiate a "Perlmutter" server (Note: Do not start a "Spin" server).

There are now several varieties of perlmutter servers. We should update that instruction to reflect the actual Jupyter servers currently available.

I would also suggest updating the release-kernel option to *e.g.* "DESI DR1 23.1" to encourage people to use that.

sbailey commented:
> > Swapping the order of software release version vs. data release version was purposeful so that optional part of the name (the data release name) is at the end instead of in the middle. Functionally they are the same.
> 
> If you actually did change the order of the inputs, that would require a change to `activate_desi_juypter.(sh|csh)`. But it looks like you changed the _formatting_ of the inputs, not the _order_. Is that what you meant?

I was referring to the order of the elements of the kernel directory name on disk and the display name, i.e. desi-edr-23.1 (old) vs. desi-23.1-edr (new).  You are correct that the order of the arguments passed to activate_desi_jupyter.sh remains the same, and now the kernel directory name and display name also match that order.

I'll also take care of updating the tutorial kernel/software installation instructions as part of the dr1 refactor.  This PR was motivated by streamlining the first steps of those instructions.

@weaverba137 I think I've addressed your comments, please re-review.

weaverba137 commented:
> I was referring to the order of the elements of the kernel directory name on disk

Ah, OK, that's fine then.

weaverba137 reviewed (approved):
Code looks good. Should we add a deprecation message to the shell script versions?

dmargala commented:
> The script is python 2.7 compatible, so it works with the ancient perlmutter default system python without requiring pre-loading a DESI-environment.

In case it's helpful, there is a python3 available on perlmutter although the version is still pretty old:

```
dmargala@perlmutter:login06:~> which python3
/usr/bin/python3
dmargala@perlmutter:login06:~> python3 --version
Python 3.6.15
```

weaverba137 commented:
> The script is python 2.7 compatible, so it works with the ancient perlmutter default system python without requiring pre-loading a DESI-environment.

I actually missed that part. If you really want to be Python 2 compatible, then you should probably add back the `from __future__ ` line.

sbailey commented:
TIL: Even without the `from __future__ import print_function` line, the script is compatible with both python 2.7 and 3.x, because python 2.7 still accepts `print(blat)` as equivalent to `print blah`.  I had originally added it thinking it was necessary to support python 2.7, but apparently the `from __future__ ...` import is only needed if using extra features of the Python 3.x function, e.g. `print(blah, end='')`.  I have explicitly tested the script with both python 2.7 and 3.10.14 .

@dmargala thanks for the note about system python3 existing; for this case we don't need that, but I'll keep that in mind if we have other cases of actually needing python3 but not wanting to require modules to be pre-loaded.

@weaverba137 good idea about the deprecation notice.  I added that to `install_jupyter_kernel.sh` and propose to leave that script in place for awhile until we're sure that we've updated any documentation and institutional memory that may be referring to it.
